docs/dev_plan.md

```diff
@@ -0,0 +1,323 @@
+input and output files should be handled by a third-party library. Similarly,
+visualization of results, derivations of statistics, and math portability
+are other areas where the ecosystem of open source scientific software
+should be leveraged, and preference should be given to software internal to
+NREL or funded by WETO.
+
+### Modular
+
+The code architecture should be structured such that each portion of code
+is responsible for a minimal unit of work. Low-level units can be combined
+at an intermediate level to do more work, but the scope of each unit
+throughout the architecture hierarchy should be explicitly described.
+This design should ensure a modular architecture, and a test of success
+is to measure the difficulty and depth of changes required to swap a
+particular unit of the code.
+
+The graphic below depicts a generic but typical data pipeline in scientific
+software. A well encapsulated and modular architecture should support
+something like swapping the library for handling YAML input files or
+changing a solver type without modifying the modules around it.
+The flow of data should be considered in discrete steps in a pipeline
+rather than a monolithic system working on the data.
+
+```{mermaid}
+flowchart LR
+
+    Start(( ))
+
+    subgraph I/O
+        direction LR
+        io1{{YAML}}
+        io2{{JSON}}
+    end
+
+    DataModel[[Data Model]]
+
+    subgraph Solver
+        direction LR
+        solver1{{Type 1}}
+        solver2{{Type 2}}
+        solver3{{Type 3}}
+    end
+
+    Output[[Output Model]]
+
+    subgraph Export
+        direction LR
+        export1{{Visualization}}
+        export2{{ASCII}}
+        export3{{Commercial}}
+    end
+
+    Finish(( ))
+
+    Start --- I/O
+    I/O --- DataModel
+    DataModel --- Solver
+    Solver --- Output
+    Output --- Export
+    Export --- Finish
+```
+
+It is critical to the sustainability and stability of `OpenTurbine`
+to maintain independence from external software even though there
+will be reliance on existing libraries for common tasks. The modular
+design should include data structures and API's that are general enough
+to support integration of third party libraries as well as the ability
+to change any library for another that accomplishes a similar task
+even if by alternative methods.
+
+### Performance
+
+A key design consideration of `OpenTurbine` is computational efficiency
+or performance. Both the quantity of work and the efficiency of data should
+be considered and measured (i.e. profiled) during any development effort.
+A modular architecture should support offloading computationally expensive
+tasks to hardware accelerators or specialized libraries, and support
+multiple options for doing so depending on user configurations and
+the computational environment.
+
+Similar to modularity in the architecure, expensive tasks should be
+structured in a kernel form. This low-level design pattern combines
```

Comment:
Just wondering if it might be appropriate here to refer to an external link that describes the kernel form with example?